File: 2020/2020W39/preppindataw39.R
library(dplyr)
library(readxl)
library(tidyr)
library(stringr)
library(splitstackshape)

final <- read_excel("F:/Data/2020W39 Input.xlsx", sheet = "Orders") %>%
  pivot_longer(.,cols=-'Person', names_to='Weekday', values_to='Order Item' ,values_drop_na=TRUE) %>%
  mutate('Order Item' = str_replace_all(`Order Item`,'\\s+with\\s+',',')) %>%
  cSplit(., 'Order Item', ',') %>%
  pivot_longer(.,cols=starts_with('Order Item'), names_to='ToDrop', values_to='Order Item' ,values_drop_na=TRUE) %>%
  mutate('Item Key' = str_remove_all(tolower(`Order Item`),'\\s*(tea|(fruit )*smoothie)$')) %>%
  merge(., read_excel("F:/Data/2020W39 Input.xlsx", sheet = "Price List") %>%
          pivot_longer(.,cols=-ends_with('Price'), names_to='Type', values_to='Item' ,values_drop_na=TRUE) %>%
          pivot_longer(.,cols=ends_with('Price'), names_to='Price Type', values_to='Price' ,values_drop_na=TRUE) %>%
          filter(`Price Type`==paste(`Type`,'Price')) %>%
          mutate('Item Key' = str_remove_all(tolower(`Item`),'\\s*(tea|(fruit )*smoothie)$')),
        by='Item Key') %>%
  group_by(`Person`) %>%
  summarise('Monthly Spend' = sum(`Price`)*4,
            'Potential Savings' = `Monthly Spend`-20,
            'Worthwhile?' = (`Potential Savings`>=0))

View(final)

File: 2020/2020W39/PreppindataW39.py
import pandas as pd
import re

order = pd.read_excel(pd.ExcelFile("F:/Data/2020W39 Input.xlsx"),'Orders')  
order = order.melt(id_vars='Person', value_name = 'Order Item', var_name='Weekday').dropna()
order = pd.concat([order[['Person','Weekday']].reset_index(drop=True), 
                   pd.DataFrame([map(str.strip, x) for x in order['Order Item'].str.split(',|with').values.tolist()])],
                  axis=1, sort=False)
order =  order.melt(id_vars=['Person','Weekday'], value_name = 'Order Item', var_name='ToDrop').dropna()
order['Order Item'] = order['Order Item'].apply(lambda x: re.sub('\s*(tea|(fruit )*smoothie)$','',x.lower()))

price = pd.read_excel(pd.ExcelFile("F:/Data/2020W39 Input.xlsx"),'Price List')  
price = price.melt(id_vars = [c for c in price.columns if re.search('Price$',c)], value_name = 'Item', var_name='Type').dropna(subset=['Item'])
price = price.melt(id_vars = ['Type','Item'], value_name = 'Price', var_name='Price Type').dropna()
price = price[price['Type'] +' Price' == price['Price Type']].copy()
price['Item'] = price['Item'].apply(lambda x: re.sub('\s*(tea|(fruit )*smoothie)$','',x.lower()))

final = pd.merge(order,price,left_on='Order Item', right_on='Item')
final = final.groupby('Person', as_index=False).agg({'Price':'sum'})
final['Monthly Spend'] = final['Price']*4
final['Potential Savings'] = final['Monthly Spend']-20
final['Worthwhile?'] = final['Potential Savings'].apply(lambda x: x>=0)
final = final.drop('Price', axis=1)
